### ISSUE TITLE
S3 storage for content (images, downloads, ...)

### ISSUE THREAD

**Issue by atmaniak** [PrestaShop/PrestaShop#29486]
### Prerequisites

- [X] I understand and accept the project's [code of conduct](https://github.com/PrestaShop/PrestaShop/blob/develop/CODE_OF_CONDUCT.md)
- [X] I have already [searched in existing features request](https://github.com/PrestaShop/PrestaShop/issues?q=is%3Aissue+label%3AImprovement%2CFeature%2CDeveloper+Feature+) and found no previous suggestion of this feature.

### Describe the solution you'd like

In order to properly host Prestashop in cloud environment, it would be nice to have an external storage like S3 supported out-of-the-box.

### Alternatives you've considered

- Using s3fs to mount image directory as normal filesystem

### Do you plan to work on this feature?

- [ ] I'm willing to contribute a [formal specification](https://github.com/PrestaShop/prestashop-specs).
- [ ] I'm willing to provide any wireframes or design assets required for this feature.
- [ ] I'm willing to submit a Pull Request that implements this feature.
- [X] I'm willing to help verify that the implemented feature works as intended and produces no unintended side effects.

**Comment by AureRita** [comment]
Hi @atmaniak  ,

Thank you for your suggestion. The Maintainer Team will take it into consideration for future developments. 

Please be aware that there is no guarantee that this feature will be developed anytime soon. If this is important to you, we strongly suggest you consider [contacting a professional](https://www.prestashop-project.org/support/) to help you create it. 

If you or anyone else is willing to [contribute this feature](https://devdocs.prestashop.com/1.7/contribute/contribute-pull-requests/) to the PrestaShop Project, the teams will be happy to help by ensuring that it fits well within the project. Please let us know if that is the case!

Thank you

**Comment by rblaurin** [comment]
This feature is incredibly important.
I've already implemented it in my company for images, css and js files, but Prestashop hasn’t made it easy. It required a lot of overrides and even some modifications to the core files.

Some key points that need to be analyzed:

- Create a wrapper class for all file operations (read, write, and delete) in PrestaShop.
- The functions could include a hook to allow the behavior of these operations to be modified, enabling a module to easily integrate these files with external storage services like AWS S3, Azure Cloud Storage, Google Cloud Storage, etc.
- All file operations in the core and modules should exclusively use this class for file handling.
- Leave the rest to the module developers...
- The modules currently available for this purpose struggle to deliver good results due to the complexity of implementation and the need for core changes and numerous overrides.

**Comment by EdwinBetanc0urt** [comment]
Is this implemented in the latest version (9.0.0)?

**Comment by boo-code** [comment]
Answering the 2025 question directly: no, this is not in 9.x, and the shape of what would be needed has not changed since the 2024 comment above.

Measured on 9.2.x:

- Nothing that could act as a storage abstraction is vendored. No flysystem, no AWS SDK, no S3 client of any kind.
- 65 files under `classes`, `controllers` and `src` reference one of `_PS_IMG_DIR_`, `_PS_PRODUCT_IMG_DIR_`, `_PS_CAT_IMG_DIR_`, `_PS_MANU_IMG_DIR_`, `_PS_SUPP_IMG_DIR_`, `_PS_STORE_IMG_DIR_` or `_PS_DOWNLOAD_DIR_`.
- 69 of those lines call a filesystem function with the constant on the same line: 43 `file_exists`, 8 `unlink`, 4 `scandir`, 4 `copy`, 3 `mkdir`, 3 `filesize`, 2 `rename`, 2 `move_uploaded_file`. That is a floor rather than a count, because most call sites build the path into a variable first and then act on the variable, which no single-line search can see.

`ImageManager` is sometimes mistaken for the wrapper this needs. It is not: its surface is `resize`, `cut`, `thumbnail`, `create`, `write`, `validateUpload`, `getMimeType` and so on, which is image processing on top of GD, and it reads and writes local paths directly like everything else. A storage layer would sit underneath it, not replace it.

So the wrapper described in the 2024 comment is the right first step, and the honest scope of it is those 65 files rather than one or two. The parts that would need deciding before any code: whether the abstraction covers reads as well as writes (serving product images from a remote store also touches URL generation and the image-type regeneration flow, not just file access), whether it is a PrestaShop interface or an existing library, and what happens to the paths already stored in the database and on disk during a migration.

None of that blocks a first PR that introduces the interface and moves one entity's images behind it, which would be a reviewable increment rather than an all-at-once change. But it does mean the feature cannot land as a configuration option on top of the current code, which is what "supported out-of-the-box" would suggest.
